src/java.base/share/classes/jdk/internal/foreign/abi/fallback/LibFallback.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.foreign.abi.fallback;
+
+import jdk.internal.foreign.abi.SharedUtils;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
+
+class LibFallback {
+    static final boolean SUPPORTED = tryLoadLibrary();
+
+    private static boolean tryLoadLibrary() {
+        try {
+            System.loadLibrary("fallbackLinker");
+        } catch (UnsatisfiedLinkError ule) {
+            return false;
+        }
+        init();
+        return true;
+    }
+
+    static final int DEFAULT_ABI = ffi_default_abi();
+
+    static final MemorySegment UINT8_TYPE = MemorySegment.ofAddress(ffi_type_uint8());
+    static final MemorySegment SINT8_TYPE = MemorySegment.ofAddress(ffi_type_sint8());
+    static final MemorySegment UINT16_TYPE = MemorySegment.ofAddress(ffi_type_uint16());
+    static final MemorySegment SINT16_TYPE = MemorySegment.ofAddress(ffi_type_sint16());
+    static final MemorySegment SINT32_TYPE = MemorySegment.ofAddress(ffi_type_sint32());
+    static final MemorySegment SINT64_TYPE = MemorySegment.ofAddress(ffi_type_sint64());
+    static final MemorySegment FLOAT_TYPE = MemorySegment.ofAddress(ffi_type_float());
+    static final MemorySegment DOUBLE_TYPE = MemorySegment.ofAddress(ffi_type_double());
+    static final MemorySegment POINTER_TYPE = MemorySegment.ofAddress(ffi_type_pointer());
+
+    static final MemorySegment VOID_TYPE = MemorySegment.ofAddress(ffi_type_void());
+    static final short STRUCT_TAG = ffi_type_struct();
+    private static final long SIZEOF_CIF = sizeofCif();
```

Comment:
Won’t calling `ffi_default_abi()` (and the other `ffi_*()` methods) throw `UnsatisfiedLinkError` if `SUPPORTED` is `false`?

Reply:
You're right. Fix here: https://github.com/minborg/jdk/pull/1